### PR TITLE
FC-1060 - Feature/tx refactor

### PR DESCRIPTION
**Pull request by bplatz**
Relies on code in same branch on fluree/db.

This transaction refactor substantially speeds up large transactions, breaks out logic into different namespaces, streamlines logic, and provides the framework for allowing different transaction types (JSON-LD, n-quads, etc).

Also adds feature for FC-1015 where new _predicate/retractDuplicates boolean predicate will force retractions when duplicates are present in a new transaction.

### CODE REVIEW

File: src/fluree/db/ledger/transact/tags.clj [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+(defn resolve
+  "Returns the subject id of the tag if it exists, or a tempid for a new tag."
+  [tag pred-info {:keys [tags db-root] :as tx-state}]
+  (go-try
+    (let [pred-name (or (pred-info :name)
+                        (throw (ex-info (str "Trying to resolve predicate name for tag resolution but name is unknown for pred: " (pred-info :id))
+                                        {:status 400
+                                         :error  :db/invalid-tx
+                                         :tags   tag})))
+          tag-name  (if (.contains tag "/") tag (str pred-name ":" tag))
```
Comment (zonotope) [inline src/fluree/db/ledger/transact/tags.clj:52]: minor, but i think you can avoid interop by using `clojure.string/includes?`

File: src/fluree/db/ledger/transact/tempid.clj [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  (flake/->Flake t const/$_tx:tempids (json/stringify tempids-map) t true nil))
+
+(defn assign-subject-ids
+  "Assigns any unresolved tempids with a permanent subject id."
+  [{:keys [tempids tempids-ordered upserts db-before t] :as tx-state} tx]
+  (try
+    (let [ecount      (assoc (:ecount db-before) -1 t)      ;; make sure to set current _tx ecount to 't' value, even if no tempids in transaction
+          tempids-map @tempids]
+      (loop [[tempid & r] @tempids-ordered
+             ;[[tempid resolved-id] & r] @tempids
```
Comment (zonotope) [inline src/fluree/db/ledger/transact/tempid.clj:106]: looks like an errant comment here

File: src/fluree/db/ledger/transact/tempid.clj [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+             ecount*      ecount]
+        (if (nil? tempid)                                   ;; finished
+          (do (reset! tempids tempids-map*)
+              (when-not (empty? upserts*)
+                (reset! upserts upserts*))
+              ;; return tx-state, don't need to update ecount in db-after, as dbproto/-with will update it
+              tx-state)
+          (if (nil? (get tempids-map tempid))
+            (let [cid      (dbproto/-c-prop db-before :id (:collection tempid))
+                  next-id  (if (= -1 cid)
```
Comment (zonotope) [inline src/fluree/db/ledger/transact/tempid.clj:118]: do we have a constant or mapping somewhere that makes it explicit what `-1` signifies as a cid? if so, i think we should use that here instead of the raw number.

File: src/fluree/db/ledger/transact/validation.clj [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+                   (do (build-function fn-subjects db p fn-type)
+                       (assoc fn-cache fn-subjects p)))))
+        ;; return whatever promise was in the cache - either one we just created or existing one if a race condition existed
+        (get-in @validate-fn-atom [:cache fn-subjects]))))
+
+
+(defn- async-response-wrapper
+  "Wraps an async response with a formatted response."
+  [async-response function]
+  (async/go
```
Comment (zonotope) [inline src/fluree/db/ledger/transact/validation.clj:94]: i think this can be:

```clojure
(defn- async-response-wrapper
  [async-response function]
  (async/pipe async-response
              (chan 1 (map function)))
```

File: src/fluree/db/ledger/transact/json.clj [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+      flakes)
+
+    ;; new and retraction flakes are different
+    :else
+    (cond-> (conj flakes new-flake retract-flake)
+            (pred-info :spec) (tx-validate/queue-pred-spec new-flake pred-info tx-state)
+            (pred-info :txSpec) (tx-validate/queue-predicate-tx-spec [new-flake retract-flake] pred-info tx-state))))
+
+
+(defn generate-statements
```
Comment (zonotope) [inline src/fluree/db/ledger/transact/json.clj:281]: Maybe not specifically for this pr and for discussion instead as this is more subjective, but I find functions this big with lots of nested loops really hard to follow. 

It's helpful for my understanding when I come across code I haven't seen before to find one main function that does (almost) nothing but call a bunch of helpers one after another, and possibly stringing those function calls together with `->` or `->>`. 

This provides a few advantages in my opinion:
- the helper function names can serve as free, lightweight documentation if they are named well. I can figure out how everything works at a high level by glancing at the "main" function, looking at the function names, and looking at the order they're called in 
- I can dive into the explicit details by focusing on individual helper functions at a time without unrelated code to distract me because that code is in different function bodies. 
- Each helper function provides even more documentation opportunities because they all have individual docstrings. 

Again, I don't have a hard blocker for this specific pr, but what do you think about that pattern?

File: src/fluree/db/ledger/bootstrap.clj [hunk 6]
```diff
@@ -98,16 +98,17 @@
...
-                     ;const/$_shard      (flake/->sid const/$_shard 999)
+                     const/$_tag        (flake/->sid const/$_tag 1000)
+                     const/$_fn         (flake/->sid const/$_fn 1000)
+                     const/$_user       (flake/->sid const/$_user 1000)
+                     const/$_auth       (flake/->sid const/$_auth 1000)
+                     const/$_role       (flake/->sid const/$_role 1000)
+                     const/$_rule       (flake/->sid const/$_rule 1000)
+                     const/$_setting    (flake/->sid const/$_setting 1000)
+                     const/$_prefix     (flake/->sid const/$_prefix 1000)
+                     const/$_shard      (flake/->sid const/$_shard 1000)})
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/ledger/bootstrap.clj:111]: Right now this is read-only and isn't being depended on. If it will be depended on we'll have to do a data upgrade, which can likely be handled with an extra block being added.

File: src/fluree/db/peer/messages.clj [hunk 7]
```diff
@@ -27,7 +27,7 @@
   [system {:keys [cmd sig] :as signed-cmd}]
   (when-not (and (string? cmd) (string? sig))
     (throw-invalid-command (str "Command map requires keys of 'cmd' and 'sig', with a json string command map and signature of the command map respectively. Provided: " (pr-str signed-cmd))))
-  (when (> (count cmd) 2000000)
+  (when (> (count cmd) 10000000)
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/peer/messages.clj:30]: We have create the _setting collection to handed ledger-level settings which I think this should be part of. We don't currently use it, but the framework is there. I agree this and some other settings should be included there.

File: test/fluree/db/ledger/docs/schema/collections.clj [hunk 8]
```diff
@@ -64,8 +64,9 @@
                                                   {:timeout 120000}))
         attemptToUpsertRes  (async/<!! (fdb/transact-async (basic/get-conn) test/ledger-chat txn))]
 
-    (is (str/includes? res "Predicate _collection/name does not allow upsert"))
-    (is (str/includes? res "duplicates an existing _collection/name (_user)"))
+    ;(is (str/includes? res "Predicate _collection/name does not allow upsert"))
+    ;(is (str/includes? res "duplicates an existing _collection/name (_user)"))
```
[the remark this answers is not in the document]
Comment (bplatz) [reply test/fluree/db/ledger/docs/schema/collections.clj:68]: Missed that.

File: dev/logback.xml [hunk 9]
```diff
@@ -15,10 +15,10 @@
 
     <logger name="user" level="ALL" />
     <logger name="ch.qos.logback" level="WARN"/>
-    <logger name="fluree.db" level="DEBUG" />
-    <logger name="fluree.db.consensus" level="DEBUG" />
+<!--    <logger name="fluree.db" level="DEBUG" />-->
```
[the remark this answers is not in the document]
Comment (bplatz) [reply dev/logback.xml:18]: Yes - someone had turned on debug logging by default in dev/repl mode. I think this should be something we turn on if we need the extra logging.

File: test/fluree/db/ledger/transact/core_test.clj [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            [fluree.db.api :as fdb]
+            [fluree.db.util.json :as json])
+  (:import (java.time Instant)))
+
+;; test these things:
+;; tag is properly resolved (add a new _predicate - the 'type' is a tag and can be compared with its repective const/$... value
+;; - if unique predicate resolves to a different existing subject, allow transaction to happen only if a different part of tx retracts existing flake
+;; - tempids - add correctly, multiple identical new create just one flake
+;; - incoming transaction cannot have a negative value? - if we allow, need to validate for prior tx, and doesn't overwrite hash, etc.
+;; dependent transactions look like it may not work correct - revisit
```
[the remark this answers is not in the document]
Comment (bplatz) [reply test/fluree/db/ledger/transact/core_test.clj:15]: Good suggestion!

File: test/fluree/db/ledger/transact/core_test.clj [hunk 11]
```diff
@@ -0,0 +1,86 @@
+(ns fluree.db.ledger.transact.core-test
```
[the remark this answers is not in the document]
Comment (bplatz) [reply test/fluree/db/ledger/transact/core_test.clj:1]: This is a new test namespace I recently added and was not part of the test suite... so probably shouldn't have ever been committed but it accidentally became part of the code base. I intend to add some of these tests, just not done yet.

File: src/fluree/db/ledger/transact/auth.clj [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+(defn add-auth-ids-permissions
+  "Figures out transaction permissions, returns map."
+  [db tx-map]
+  (go-try
+    (let [{:keys [auth authority]} tx-map
+          auth-id-ch      (dbproto/-subid db ["_auth/id" auth] true)
+          authority-id-ch (when authority
+                            (let [authority-id (if (string? authority) ["_auth/id" authority] authority)]
+                              (dbproto/-subid db authority-id true)))
+          auth-sid        (async/<! auth-id-ch)
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/ledger/transact/auth.clj:26]: Was that way to perform both searches in parallel, but I'll refactor a bit to still accomplish the parallelism but also be a bit more clean.

File: src/fluree/db/ledger/transact/error.clj [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+  but we still want to record the tx. We'll assume cmd-data is a string
+  and instead of extracting a command out of it (which didn't work"
+  [e db cmd-data t]
+  (try
+    (let [{:keys [message status error]} (decode-exception e)
+          {:keys [sig cmd]} (:command cmd-data)
+          txid             (crypto/sha3-256 cmd)
+          tx-state         {:tx-string cmd :signature sig :t t :txid txid :fuel (atom {:spent 0})}
+          flakes           (->> (tx-meta/tx-meta-flakes tx-state message)
+                                (into (flake/sorted-set-by flake/cmp-flakes-block)))
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/ledger/transact/error.clj:110]: Agree, I have this resolved in uncommitted work in the feature/prefix branch where I've refactored errors a bit, so would prefer to just address it once that PR lands.

File: src/fluree/db/ledger/transact/json.clj [hunk 14]
```diff
@@ -0,0 +1,648 @@
...
+  [predicate-value tx-state]
+  (cond (txi-list? predicate-value)
+        (let [txis (map #(assoc % :_id (tempid/new (:_id %) tx-state)) predicate-value)]
+          [(map :_id txis) txis])
+
+        (txi? predicate-value)
+        (let [tempid (tempid/new (:_id predicate-value) tx-state)]
+          [tempid (assoc predicate-value :_id tempid)])
+
+        :else nil))
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/ledger/transact/json.clj:60]: OK

File: src/fluree/db/ledger/transact/json.clj [hunk 15]
```diff
@@ -0,0 +1,648 @@
...
+(defn- resolve-nested-txi
+  "Takes a predicate-value from a transaction item (what will be the Flakes' .-o value),
+  and if that value contains children/nested transaction item(s), returns a two-tuple of
+  [tempid(s) nested-txi(s)-with-updated-tempids], else returns nil if no nested txi.
+
+   Tempids need to be validated and generated here, because if they are not unique tempids
+   we must make them unique so they point to the correct subjects once flattened."
+  [predicate-value tx-state]
+  (cond (txi-list? predicate-value)
+        (let [txis (map #(assoc % :_id (tempid/new (:_id %) tx-state)) predicate-value)]
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/ledger/transact/json.clj:53]: That won't work here... it isn't taking the full argument, just taking (:_id %)

File: src/fluree/db/ledger/transact/txfunction.clj [hunk 16]
```diff
@@ -0,0 +1,18 @@
+(ns fluree.db.ledger.transact.txfunction
+  (:require [fluree.db.dbfunctions.core :as dbfunctions]))
+
+;; functions related to transaction functions
+
+;; TODO - can probably parse function string to final 'lisp form' when generating TxFunction
+(defrecord TxFunction [fn-str])
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/ledger/transact/txfunction.clj:7]: It allows for polymorphism without recasting all values as a map so we can embed a type.  I suppose it could be accomplished in different ways - but this seemed efficient (we don't recast all values, just tx-functions), and keeps logic clear downstream as we can simply do (tx-function? val). I was thinking we'd have other values we'd cast as types too, just started here.

File: src/fluree/db/ledger/transact/tags.clj (hunk 1 above)
Comment (bplatz) [reply src/fluree/db/ledger/transact/tags.clj:52]: Great.. fixing.

File: src/fluree/db/ledger/transact/tempid.clj (hunk 3 above)
Comment (bplatz) [reply src/fluree/db/ledger/transact/tempid.clj:118]: Good idea, adding it to fluree/db constants

File: src/fluree/db/ledger/transact/json.clj (hunk 5 above)
Comment (bplatz) [reply src/fluree/db/ledger/transact/json.clj:281]: Completely agree, and where I could I did that throughout this PR... this is the one sore spot. I think I have this one broken up in ~half in the work on feature/prefix, so I don't expect it to last long in this exact state.

While this ballooned a little bit over the refactor, part of the reason this wasn't broken up farther (it is about half the size of the original) is there are a lot of conditional async chans. By wrapping them many functions, it results in creating additional channels that are not needed. This is the core processor, so a large transaction this could mean 100K+ extra async channels that were not needed. I'm concerned about wrapping that many values around channels that never needed to be wrapped, and how that might impact performance in this very performance-critical part of the code.

File: src/fluree/db/ledger/transact/validation.clj (hunk 4 above)
Comment (bplatz) [reply src/fluree/db/ledger/transact/validation.clj:94]: Cool alternative, putting in place.

File: src/fluree/db/ledger/transact/tempid.clj (hunk 2 above)
Comment (bplatz) [reply src/fluree/db/ledger/transact/tempid.clj:106]: ✔️

File: src/fluree/db/ledger/transact/retract.clj [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+(defn retract-components
+  "Checks flakes to see if any are a component, and if so, finds additional retractions and returns."
+  [flakes {:keys [db-root] :as tx-state}]
+  (go-try
+    (loop [[^Flake flake & r] flakes
+           components #{}]
+      (if (nil? flake)
+        components
+        (let [component? (true? (dbproto/-p-prop db-root :component (.-p flake)))]
+          (if component?
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/ledger/transact/retract.clj:23]: Resolving by implementing a helper function (component-flake? ...) to make this cleaner, but retain meaning.

File: src/fluree/db/ledger/transact/json.clj [hunk 18]
```diff
@@ -0,0 +1,648 @@
...
+
+(defn register-unique!
+  "Registers unique value in the tx-state and return true if successful.
+  Will be unsuccessful if the ident already exists in the unique value cache
+  and return false.
+
+  Ident is a two-tuple of [pred-id object/value]"
+  [ident {:keys [uniques] :as tx-state}]
+  (if (contains? @uniques ident)
+    false                                                   ;; already registered, should throw downstream
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/ledger/transact/json.clj:139]: Nice catch, fixed in next commit.

File: src/fluree/db/ledger/transact/retract.clj [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+                       (<? (query-range/index-range db-root :spot = [subject-id predicate-id]))
+                       (<? (query-range/index-range db-root :spot = [subject-id predicate-id object])))
+          components (when (dbproto/-p-prop db-root :component predicate-id)
+                       (<? (retract-components flakes tx-state)))]
+      (->> flakes
+           (map #(flake/flip-flake % t))
+           (into components)))))
+
+
+;; TODO - below, instead of async/into,could use a transducer to return a single clean channel that concats, and not need to use go-try here
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/ledger/transact/retract.clj:59]: I'd need an example here. The only way I could make this work ends up with a less clear function.

[the remark this answers is not in the document]
Comment (zonotope) [reply src/fluree/db/ledger/transact/retract.clj:67]: What I was thinking of doesn't get rid of the `async/into` either, but It does get rid of the `go` though:

```clojure
(let [out (chan 1 (map (partial apply concat)))]
  (->> objects
       (map #(flake subject-id predicate-id % tx-state))
       async/merge
       (async/into [])
       (async/pipe out))
```
This isn't that different from how it is now in any case. Take it or leave it.

**Review by zonotope — APPROVED** [review]

**Comment by bplatz** [comment]
> @bplatz Noticed a tx-refactor-specific issue when trying to delete all values on a multi predicate. For example a `_user` with `/favNums` of `[1, 2, 3]`. If you issue this txn to delete all favNums:
> 
> `[{ _id: ["_user/username", "jdoe"], favNums: null }]`
> 
> It returns a `"Don't know how to create ISeq from: clojure.lang.Keyword"`

@aaj3f thanks for the catch, I addressed this
